Micronaut v4: `IllegalStateException: cannot send more responses than requests`

### Expected Behavior

1. upgrading to micronaut v4 should not yield new errors for regular http requests
2. we upgraded to micronaut 4 and observed the following exception

### Actual Behaviour

our app happens to use expect/continue, but these exceptions surface even if this functionality is disabled in relevant clients, so it seems not related:

```
[16:25:03.820 ERROR] io.micronaut.http.server.netty.RoutingInBoundHandler Micronaut Server Error - No request state present. Cause: java.lang.IllegalStateException: cannot send more responses than requests
io.netty.handler.codec.EncoderException: java.lang.IllegalStateException: cannot send more responses than requests
        at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
        at io.netty.handler.codec.MessageToMessageCodec.write(MessageToMessageCodec.java:116)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856)
        at io.netty.channel.ChannelDuplexHandler.write(ChannelDuplexHandler.java:115)
        at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.write(WebSocketServerExtensionHandler.java:124)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856)
        at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:302)
        at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:131)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:923)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:941)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:966)
        at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:934)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler.write(PipeliningServerHandler.java:257)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler$FullOutboundHandler.writeSome(PipeliningServerHandler.java:765)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler.writeSome(PipeliningServerHandler.java:281)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler.channelWritabilityChanged(PipeliningServerHandler.java:226)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:528)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:505)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:498)
        at io.netty.handler.stream.ChunkedWriteHandler.channelWritabilityChanged(ChunkedWriteHandler.java:146)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:526)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:505)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:498)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelWritabilityChanged(CombinedChannelDuplexHandler.java:448)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelWritabilityChanged(ChannelInboundHandlerAdapter.java:129)
        at io.netty.channel.CombinedChannelDuplexHandler.channelWritabilityChanged(CombinedChannelDuplexHandler.java:271)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:526)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:505)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelWritabilityChanged(AbstractChannelHandlerContext.java:498)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelWritabilityChanged(DefaultChannelPipeline.java:1433)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:524)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelWritabilityChanged(AbstractChannelHandlerContext.java:505)
        at io.netty.channel.DefaultChannelPipeline.fireChannelWritabilityChanged(DefaultChannelPipeline.java:931)
        at io.netty.channel.ChannelOutboundBuffer.fireChannelWritabilityChanged(ChannelOutboundBuffer.java:638)
        at io.netty.channel.ChannelOutboundBuffer.setWritable(ChannelOutboundBuffer.java:604)
        at io.netty.channel.ChannelOutboundBuffer.decrementPendingOutboundBytes(ChannelOutboundBuffer.java:197)
        at io.netty.channel.ChannelOutboundBuffer.remove(ChannelOutboundBuffer.java:283)
        at io.netty.channel.ChannelOutboundBuffer.removeBytes(ChannelOutboundBuffer.java:362)
        at io.netty.channel.socket.nio.NioSocketChannel.doWrite(NioSocketChannel.java:438)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:931)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.flush0(AbstractNioChannel.java:354)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.flush(AbstractChannel.java:895)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.flush(DefaultChannelPipeline.java:1372)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:921)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:907)
        at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:893)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.flush(CombinedChannelDuplexHandler.java:531)
        at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:125)
        at io.netty.channel.CombinedChannelDuplexHandler.flush(CombinedChannelDuplexHandler.java:356)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:923)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:907)
        at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:893)
        at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:313)
        at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:131)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:923)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:941)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:966)
        at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:934)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler.write(PipeliningServerHandler.java:257)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler$FullOutboundHandler.writeSome(PipeliningServerHandler.java:765)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler.writeSome(PipeliningServerHandler.java:281)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler$OutboundAccess.write(PipeliningServerHandler.java:633)
        at io.micronaut.http.server.netty.handler.PipeliningServerHandler$OutboundAccess.lambda$write$0(PipeliningServerHandler.java:623)
        at io.micronaut.configuration.metrics.binder.netty.MonitoredQueue$TimedRunnable.run(MonitoredQueue.java:263)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1623)
Caused by: java.lang.IllegalStateException: cannot send more responses than requests
        at io.netty.handler.codec.http.HttpContentEncoder.encode(HttpContentEncoder.java:125)
        at io.micronaut.http.server.netty.SmartHttpContentCompressor.encode(SmartHttpContentCompressor.java:65)
        at io.micronaut.http.server.netty.SmartHttpContentCompressor.encode(SmartHttpContentCompressor.java:33)
        at io.netty.handler.codec.MessageToMessageCodec$1.encode(MessageToMessageCodec.java:67)
        at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:90)
        ... 77 common frames omitted
```

### Steps To Reproduce

investigating

### Environment Information

- Operating System: macOS Ventura 13.3.1
- JDK:
```
openjdk version "20.0.1" 2023-04-18
OpenJDK Runtime Environment Zulu20.30+11-CA (build 20.0.1+9)
OpenJDK 64-Bit Server VM Zulu20.30+11-CA (build 20.0.1+9, mixed mode, sharing)
```

### Example Application

pending

### Version

4.0.0-Mx

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/handler/PipeliningServerHandler.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/handler/PipeliningServerHandler.java
@@ -131,6 +131,10 @@ public final class PipeliningServerHandler extends ChannelInboundHandlerAdapter 
      * {@code true} iff we should flush on {@link #channelReadComplete}.
      */
     private boolean flushPending = false;
+    /**
+     * {@code true} inside {@link #writeSome()} to avoid reentrancy.
+     */
+    private boolean writing = false;
 
     public PipeliningServerHandler(RequestHandler requestHandler) {
         this.requestHandler = requestHandler;
@@ -265,23 +269,32 @@ public final class PipeliningServerHandler extends ChannelInboundHandlerAdapter 
      * Write some data if possible.
      */
     private void writeSome() {
-        while (ctx.channel().isWritable()) {
-            // if we have no outboundHandler, check whether the first queued response is ready
-            if (outboundHandler == null) {
-                OutboundAccess next = outboundQueue.peek();
-                if (next != null && next.handler != null) {
-                    outboundQueue.poll();
-                    outboundHandler = next.handler;
-                } else {
-                    return;
+        if (writing) {
+            // already inside writeSome
+            return;
+        }
+        writing = true;
+        try {
+            while (ctx.channel().isWritable()) {
+                // if we have no outboundHandler, check whether the first queued response is ready
+                if (outboundHandler == null) {
+                    OutboundAccess next = outboundQueue.peek();
+                    if (next != null && next.handler != null) {
+                        outboundQueue.poll();
+                        outboundHandler = next.handler;
+                    } else {
+                        return;
+                    }
+                }
+                OutboundHandler oldHandler = outboundHandler;
+                oldHandler.writeSome();
+                if (outboundHandler == oldHandler) {
+                    // handler is not done yet
+                    break;
                 }
             }
-            OutboundHandler oldHandler = outboundHandler;
-            oldHandler.writeSome();
-            if (outboundHandler == oldHandler) {
-                // handler is not done yet
-                break;
-            }
+        } finally {
+            writing = false;
         }
     }
 

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/PipeliningServerHandlerSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/PipeliningServerHandlerSpec.groovy
@@ -4,6 +4,7 @@ import io.micronaut.http.netty.stream.StreamedHttpRequest
 import io.netty.buffer.Unpooled
 import io.netty.channel.ChannelHandlerContext
 import io.netty.channel.ChannelOutboundHandlerAdapter
+import io.netty.channel.ChannelPromise
 import io.netty.channel.embedded.EmbeddedChannel
 import io.netty.handler.codec.http.DefaultFullHttpRequest
 import io.netty.handler.codec.http.DefaultFullHttpResponse
@@ -192,6 +193,45 @@ class PipeliningServerHandlerSpec extends Specification {
         when:
         ch.writeInbound(new DefaultLastHttpContent(Unpooled.wrappedBuffer("foo".getBytes(StandardCharsets.UTF_8))))
         then:
+        ch.readOutbound() == resp
+        ch.readOutbound() == null
+    }
+
+    def 'nested write'() {
+        given:
+        def mon = new MonitorHandler()
+        def resp = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.NO_CONTENT)
+        def ch = new EmbeddedChannel(mon, new ChannelOutboundHandlerAdapter() {
+            @Override
+            void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                super.write(ctx, msg, promise)
+                ctx.fireChannelWritabilityChanged()
+            }
+        }, new PipeliningServerHandler(new RequestHandler() {
+            @Override
+            void accept(ChannelHandlerContext ctx, HttpRequest request, PipeliningServerHandler.OutboundAccess outboundAccess) {
+                assert request instanceof FullHttpRequest
+                request.release()
+                outboundAccess.writeFull(resp)
+            }
+
+            @Override
+            void handleUnboundError(Throwable cause) {
+                cause.printStackTrace()
+            }
+        }))
+
+        expect:
+        mon.read == 1
+        mon.flush == 0
+
+        when:
+        def req = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/", Unpooled.EMPTY_BUFFER)
+        ch.writeInbound(req)
+        then:
+        ch.checkException()
+        mon.read == 2
+        mon.flush == 1
         ch.readOutbound() == resp
         ch.readOutbound() == null
     }

--- a/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/PipeliningServerHandlerSpec.groovy
+++ b/http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/PipeliningServerHandlerSpec.groovy
@@ -23,6 +23,7 @@ import io.netty.handler.codec.http.HttpResponseStatus
 import io.netty.handler.codec.http.HttpVersion
 import reactor.core.publisher.Flux
 import reactor.core.publisher.Sinks
+import spock.lang.Issue
 import spock.lang.Specification
 
 import java.nio.charset.StandardCharsets
@@ -197,6 +198,7 @@ class PipeliningServerHandlerSpec extends Specification {
         ch.readOutbound() == null
     }
 
+    @Issue('https://github.com/micronaut-projects/micronaut-core/issues/9366')
     def 'nested write'() {
         given:
         def mon = new MonitorHandler()

